### PR TITLE
Update cmake rpm generation and add README for Fedora/rpm

### PR DESCRIPTION
**Pull request by Veratil**
This adds 2 new files, *Fedora.README.md* and *awesome.spec.in*, and adds a section in *Packaging.cmake* for the RPM generator.

*Fedora.README.md* is instructions for installing on Fedora (and hopefully any rpm-based system).

*awesome.spec.in* is a modified version of @thmo's Fedora spec file that I originally updated for v4.0 and then tore down again to make it work for CPack.

On the spec file, I decided to use the [Snapshot versioning](https://fedoraproject.org/wiki/Packaging:Versioning#Snapshot_packages) instead of the git string generated earlier in Packaging.cmake since rpm throws an error with `-` being in the string. I kept .dirty at the end to keep in line with `-dirty`.

There are two rpm's generated: `awesome` and `awesome-doc`. This is the same as @thmo did for Fedora, but we can easily bundle the docs into the `awesome` rpm if wanted.

### CODE REVIEW

File: Packaging.cmake [hunk 1]
```diff
@@ -62,6 +62,21 @@ set(CPACK_DEBIAN_PACKAGE_MAINTAINER "devnull@example.com")
...
 set(CPACK_RPM_EXCLUDE_FROM_AUTO_FILELIST_ADDITION "/etc/xdg;/usr/share/xsessions")
 
+if(CPACK_GENERATOR MATCHES "RPM")
+    set(CPACK_PACKAGE_FILE_NAME ${CPACK_PACKAGE_NAME})
+    execute_process(
+        COMMAND ${GIT_EXECUTABLE} log -1 --date=format:%Y%m%d --format=%cdgit%h.dirty
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE GIT_RELEASE_STRING
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
```
Comment (psychon) [inline Packaging.cmake]: This only works when building from Git (and git is installed). Should this also be made to work without Git and/or when building from a release tarball (where the version is saved in a file called .version_stamp)?

File: awesome.spec.in [hunk 2]
```diff
@@ -68,10 +68,8 @@ Requires:	cairo-gobject%{?_isa}
 Requires:	startup-notification >= 0.10
 # terminal used in the default configuration
 Requires:	xterm
-# optional but useful
-Requires:	rlwrap
-# default editor
-Requires:	vi
+# editor used in the default configuration
+Requires:	nano
```
Comment (psychon) [inline awesome.spec.in]: This change should be moved to the previous commit.

File: Fedora.README.md [hunk 3]
```diff
@@ -101,7 +101,7 @@ mock -r fedora-24-x86_64 --clean
 
 ## Notes
 
-The rpms generated follow the naming convention of `[package name][-component]-[version]-[release][dist][git string]`. `[git string]` is in the format `YYYYMMDDgitHASH` where `YYYYMMDD` is the date of the latest commit and `HASH` is the short hash id for the latest commit. This allows future rpm builds of different dates to be recognized as newer without problem.
+The rpms generated follow the naming convention of `[package name][-component]-[version]-[release].[git string][dist]`. `[git string]` is in the format `YYYYMMDDgitHASH` where `YYYYMMDD` is the date of the latest commit and `HASH` is the short hash id for the latest commit. This allows future rpm builds of different dates to be recognized as newer without problem.
```
Comment (psychon) [inline Fedora.README.md]: Perhaps this whole PR should be squashed...?

**Comment by psychon** [comment]
My problem with this is that it adds two more places where the build dependencies are listed. All of these lists *will* get out of sync. I have no idea, but hard would it be to actually test this stuff on Travis? Like, do something with docker so that the whole building-a-package is actually tested and missing new dependencies will be caught?

**Comment by thmo** [comment]
Keeping these lists in sync is indeed a problem, but something I have to do for Fedora anyway...

But the lists (even multiple times) in the readme can be removed, I think. Just mention the `dnf builddep` command which can be used on the spec file to install all BR. Also, for mock, it shouldn't be necessary to install all requirements beforehand, mock should automatically install them (or use `mock --dnf-cmd builddep ...` same way like above, when doing development inside the mock environment).

File: Packaging.cmake (hunk 1 above)
Comment (Veratil) [inline Packaging.cmake]: Yeah, this is just initial steps on getting things working. I'm still learning CMake and all.

File: Fedora.README.md (hunk 3 above)
Comment (Veratil) [inline Fedora.README.md]: I'm actually not familiar with how to do that.

**Comment by Veratil** [comment]
@thmo I wasn't able to use builddep on the spec.in file, which is why I didn't mention it. It kept giving me "Error: no package found". I wasn't able to figure out a way around it so I had to make the list. Mock couldn't do it either since it relies on the builddep command as well.

File: Fedora.README.md (hunk 3 above)
Comment (psychon) [inline Fedora.README.md]: Don't worry, we can handle that part (if in the end everything should be squashed into a single commit)

**Comment by thmo** [comment]
I am by no means a cmake expert, but what about creating a target that only configures the spec file from the spec.in? And the use that with builddep?

**Comment by Veratil** [comment]
That should be possible to do, or it could be done after configuring as well. After configure_file() is called we could call `dnf builddep awesome.spec` to gather everything.

**Comment by psychon** [comment]
While failing to fall asleep, I came up with another question: What's the point of this?
I don't think we need the rpm generated by `make package` to pass Fedora's quality standards. It should just work. So, for example, I am fine with it not listing the needed dependencies. The intended use case (in my opinion) is to install it on the machine on which it was built and so the dependencies should be there already.

Put differently: What's the added-maintenance-burden vs gain ratio here?

**Comment by Veratil** [comment]
I plan on working on this a little more when I get the time, but you're right, we don't need it to pass quality standards as long as it just works.

The problem with it though is that all the dependencies aren't clear up front, and you have to run make multiple times just to see what dependency you need next. If I recall correctly, I had to run make *5 or 6 times* before I finally managed to get it running past the dependency check. What makes it worse is that I had to deduce what package was the proper one since pkgconfig lists things differently than the actual package name. Having a list of explicit packages gives the user a clear list of what's needed.

That being said, I'm okay with refusing the PR altogether as long as we find a way to list, or at best install, the dependencies needed to build awesome.

**Comment by psychon** [comment]
So for you the main point is the README entry and not the `make package` stuff. Ok.
(Didn't fedora have a way to install packages by pkg-config name? Or do things like `pkgconfig(xcb-shape)` only work in source RPMs?)

**Comment by Veratil** [comment]
The biggest issue was the packages needed, yes.

The CPack spec generated didn't help much with that aspect either, and I believe it didn't even put in BuildRequires since CMake handles that. Hence I used Fedora's spec since it is established and working (and should for any rpm based distro).

> (Didn't fedora have a way to install packages by pkg-config name? Or do things like pkgconfig(xcb-shape) only work in source RPMs?)

Yes, rpm spec allows you do use pkgconfig([name]) for alternative packages to provide the same as others without specifying the package directly. Theoretically you can install the dependencies using `dnf builddep`, but I wasn't having luck with that since I didn't have the packages installed for pkg-config to recognize what was needed. I'm not sure why it was happening, but it seemed like a chicken and egg issue.

**Comment by thmo** [comment]
`pkgconfig(xcb-shape)` is a (virtual) dependency that can be installed by `dnf install` directly as well as via the `dnf builddep` command - as can any other type of dependency, so the `README` could as well list the dependencies in `pkgconfig(...)` style.

**Comment by Veratil** [comment]
@thmo so why wasn't it finding the packages for me? Inside mock it seemed to work find, but on my system it wasn't working the same way.

**Comment by thmo** [comment]
Well that's hard to debug remotely (and slightly off-topic btw). Try to run `dnf install 'pkgconfig(xcb-shape)` and see what it does (or not). At least on the systems that I have access to, it works.

**Comment by Veratil** [comment]
Starting (over) from the master branch, the rpm generated will fail to install from these errors on Fedora:

```
Running transaction test
Error: Transaction check error:
  file /usr/local from install of awesome-4.0.0.0~git150_g5a914b3-1.x86_64 conflicts with file from package filesystem-3.2-37.fc24.x86_64
  file /usr/local/etc from install of awesome-4.0.0.0~git150_g5a914b3-1.x86_64 conflicts with file from package filesystem-3.2-37.fc24.x86_64
  file /usr/local/share from install of awesome-4.0.0.0~git150_g5a914b3-1.x86_64 conflicts with file from package filesystem-3.2-37.fc24.x86_64
```

which is caused by these lines:

```
  1 %files
  2 %defattr(-,root,root,-)
  3 "/usr/bin/awesome"
  4 "/usr/bin/awesome-client"
  5 %dir "/usr/local"
  6 %dir "/usr/local/etc"
  7 %dir "/usr/local/etc/xdg"
  8 %dir "/usr/local/etc/xdg/awesome"
  9 %dir "/usr/local/share"
 10 %dir "/usr/local/share/awesome"
```

---

rpmbuild has a `%cmake` macro which expands to:

```
-14: cmake      
  CFLAGS="${CFLAGS:-%optflags}" ; export CFLAGS ; 
  CXXFLAGS="${CXXFLAGS:-%optflags}" ; export CXXFLAGS ; 
  FFLAGS="${FFLAGS:-%optflags%{?_fmoddir: -I%_fmoddir}}" ; export FFLAGS ; 
  FCFLAGS="${FCFLAGS:-%optflags%{?_fmoddir: -I%_fmoddir}}" ; export FCFLAGS ; 
  %{?__global_ldflags:LDFLAGS="${LDFLAGS:-%__global_ldflags}" ; export LDFLAGS ;} 
  %__cmake \
        -DCMAKE_C_FLAGS_RELEASE:STRING="-DNDEBUG" \
        -DCMAKE_CXX_FLAGS_RELEASE:STRING="-DNDEBUG" \
        -DCMAKE_Fortran_FLAGS_RELEASE:STRING="-DNDEBUG" \
        -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON \
        -DCMAKE_INSTALL_PREFIX:PATH=%{_prefix} \
        -DINCLUDE_INSTALL_DIR:PATH=%{_includedir} \
        -DLIB_INSTALL_DIR:PATH=%{_libdir} \
        -DSYSCONF_INSTALL_DIR:PATH=%{_sysconfdir} \
        -DSHARE_INSTALL_PREFIX:PATH=%{_datadir} \
%if "%{?_lib}" == "lib64" 
        %{?_cmake_lib_suffix64} \
%endif 
        -DBUILD_SHARED_LIBS:BOOL=ON
```

I think this might be a better path to attempt. Reason being is that on Fedora, for instance, it's forbidden to install anything to `/usr/local`. Granted this is a locally made and installed rpm, but keeping with system settings is the nice thing to do.

**Comment by Veratil** [comment]
Fixing the `%dir` errors is simple with:

```diff
diff --git a/Packaging.cmake b/Packaging.cmake
index b63e3bb..a42bc68 100644
--- a/Packaging.cmake
+++ b/Packaging.cmake
@@ -60,7 +60,12 @@ endif(NOT CPACK_GENERATOR)
 set(CPACK_PACKAGE_NAME "awesome")
 set(CPACK_DEBIAN_PACKAGE_MAINTAINER "devnull@example.com")
 set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "A tiling window manager")
-set(CPACK_RPM_EXCLUDE_FROM_AUTO_FILELIST_ADDITION "/etc/xdg;/usr/share/xsessions")
+set(CPACK_RPM_EXCLUDE_FROM_AUTO_FILELIST_ADDITION
+    "/etc/xdg"
+    "/usr/share/xsessions"
+    "/usr/local"
+    "/usr/local/etc"
+    "/usr/local/share")
 
 if(CPACK_GENERATOR)
     include(CPack)
```

Another problem I found is that `/etc/xdg/awesome` is not created. It is instead made at `/usr/local/etc/xdg/awesome`. The same is found with `/usr/share/xsessions/awesome.desktop`, instead installed to `/usr/local/share/xsessions/awesome.desktop`.

From the generated spec:
`%config "/usr/local/etc/xdg/awesome/rc.lua"`
`%config "/usr/local/share/xsessions/awesome.desktop"`

**Comment by psychon** [comment]
Your patch would be fine with me.

> Another problem I found is that /etc/xdg/awesome is not created. It is instead made at /usr/local/etc/xdg/awesome.

That will be fixed by #1463 (the fix is not to place it at another place, but to make awesome find it in this place).

> The same is found with /usr/share/xsessions/awesome.desktop, instead installed to /usr/local/share/xsessions/awesome.desktop

Hm. No idea what to do about this one. I would hope that "things" also look at `/usr/local` for these desktop files, but since you mention it I guess this hope is false.

**Comment by Veratil** [comment]
> I would hope that "things" also look at /usr/local for these desktop files, but since you mention it I guess this hope is false.

I don't think Fedora considers `/usr/local` to exist since it's forbidden to use per their packaging standards. I'm honestly not sure if this would work or not.

**Comment by thmo** [comment]
I am not entirely sure about gdm's search path, but I'd guess it will not find the .desktop file there.

In any case, an RPM installing things to `/usr/local` is plain wrong - for Fedora, but also in general, imho.

**Comment by psychon** [comment]
@thmo Then you need `-DCMAKE_INSTALL_PREFIX=/usr` when calling CMake. I know this is not ideal, but it is still better than installing things directly and having no way to uninstall things. And yes, the "uninstalling" is the main gain of `make package` for me.

**Comment by Veratil** [comment]
@psychon yeah that's taken care of by the `%cmake` macro in rpmbuild I posted above.

**Comment by Veratil** [comment]
Thanks for all the input everyone. I'm closing this for looking into fixing it up in a different way without having to add new files (which I think would make everyone happier).

**Pull request closed without merging**